packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -418,21 +427,23 @@ export class TokenDetectionController extends StaticIntervalPollingController<
         };
       }
     }
-    const { chainId } = this.messagingSystem.call(
-      'NetworkController:getProviderConfig',
+    const { selectedNetworkClientId } = this.messagingSystem.call(
+      'NetworkController:getState',
     );
-    const newNetworkClientId = this.messagingSystem.call(
-      'NetworkController:findNetworkClientIdByChainId',
-      chainId,
+    const {
+      configuration: { chainId },
+    } = this.messagingSystem.call(
+      'NetworkController:getNetworkClientById',
+      selectedNetworkClientId,
     );
     return {
       chainId,
-      networkClientId: newNetworkClientId,
+      networkClientId: selectedNetworkClientId,
```

Comment:
Replaces `getProviderConfig`, `findNetworkClientIdByChainId` actions with [`getState`, `getNetworkClientById`](https://github.com/MetaMask/core/pull/3938/commits/c75fb3bd37ddf2715ed250bdd2c2331a54008b72)

I used `getNetworkClientById` instead of `getNetworkConfigurationByNetworkClientId` to avoid having to use `this.#chainId` as a fallback value for the case where the returned network configuration is undefined.

Reply:
@bergeron @jiexi @adonesky1 Thoughts on this approach for retrieving `networkClientId`, `chainId`? https://github.com/MetaMask/core/pull/3938/commits/c75fb3bd37ddf2715ed250bdd2c2331a54008b72

Reply:
That approach looks okay to me, but wondering if you could get both in 1 call:

```ts
    const {
      selectedNetworkClientId,
      providerConfig: { chainId },
    } = this.messagingSystem.call('NetworkController:getState');
```

Reply:
Looks good to me as long as there's no edge case where the selected network client's `chainId` is inconsistent with `providerConfig`. @mcmire What do you think?

Reply:
Applied here: https://github.com/MetaMask/core/pull/3938/commits/d6caee2d799f72ea10bdcb11db98ce26dcb6edf6

Reply:
That makes sense. We might as well do the same thing for lines 418-428, huh?

Reply:
I'm a little confused here. For `detectTokens` to work independently of the class state, shouldn't the user be able to pass in a network client ID that's not the currently selected network's and get the corresponding chain ID? 

It seems like 418-428 is handling this case where the user-supplied network client ID is the source of truth, instead of the `selectedNetworkClientId` in NetworkState.

Otherwise, if the only network client ID ever used in `detectTokens` is the `selectedNetworkClientId` from NetworkState, shouldn't we remove the `networkClientId` params from `detectTokens`, `restartTokenDetection`, and `getCorrect...`?

Reply:
Oh, sorry. What I meant was that it looks like further up we are using `getNetworkConfigurationByNetworkClientId` to get the configuration for the given network client, then we are checking to see whether we have a configuration, then we are getting the chain ID from the configuration. But in your changes here you are using `getNetworkClientById`. So do we want to be consistent? I'm thinking this:

``` suggestion
  #getCorrectChainIdAndNetworkClientId(networkClientId?: NetworkClientId): {
    chainId: Hex;
    networkClientId: NetworkClientId;
  } {
    if (networkClientId) {
      const {
        configuration: { chainId },
      } = this.messagingSystem.call(
        'NetworkController:getNetworkClientById',
        networkClientId,
      );
      return {
        chainId,
        networkClientId,
      };
    }

    const { selectedNetworkClientId } = this.messagingSystem.call(
      'NetworkController:getState',
    );
    const {
      configuration: { chainId },
    } = this.messagingSystem.call(
      'NetworkController:getNetworkClientById',
      selectedNetworkClientId,
    );
    return {
      chainId,
      networkClientId: selectedNetworkClientId,
    };
  }
```

And if we can do this, then we can flip it around:

``` suggestion
  #getCorrectChainIdAndNetworkClientId(givenNetworkClientId?: NetworkClientId): {
    chainId: Hex;
    networkClientId: NetworkClientId;
  } {
    const networkClientId =
      givenNetworkClientId ??
      this.messagingSystem.call(
        'NetworkController:getState',
      ).selectedNetworkClientId;
    const {
      configuration: { chainId },
    } = this.messagingSystem.call(
      'NetworkController:getNetworkClientById',
      networkClientId,
    );
    return { chainId, networkClientId };
  }
```

Reply:
Oh I see what you're saying. Okay, that makes sense!